examples/tool_chat_template_hermes_2_pro.jinja

```diff
@@ -0,0 +1,120 @@
+    {{- '"description": "' + tool.name + '(' }}
+    {%- for param_name, param_fields in tool.parameters.properties|items %}
+        {{- param_name + ": " + json_to_python_type(param_fields) }}
+        {%- if not loop.last %}
+            {{- ", " }}
+        {%- endif %}
+    {%- endfor %}
+    {{- ")" }}
+    {%- if tool.return is defined %}
+        {{- " -> " + json_to_python_type(tool.return) }}
+    {%- endif %}
+    {{- " - " + tool.description + "\n\n" }}
+    {%- for param_name, param_fields in tool.parameters.properties|items %}
+        {%- if loop.first %}
+            {{- "    Args:\n" }}
+        {%- endif %}
+        {{- "        " + param_name + "(" + json_to_python_type(param_fields) + "): " + param_fields.description|trim }}
+    {%- endfor %}
+    {%- if tool.return is defined and tool.return.description is defined %}
+        {{- "\n    Returns:\n        " + tool.return.description }}
+    {%- endif %}
+    {{- '"' }}
+    {{- ', "parameters": ' }}
+    {%- if tool.parameters.properties | length == 0 %}
+        {{- "{}" }}
+    {%- else %}
+        {{- tool.parameters|tojson }}
+    {%- endif %}
+    {{- "}" }}
+    {%- if not loop.last %}
+        {{- "\n" }}
+    {%- endif %}
+{%- endfor %}
+{{- " </tools>" }}
+{{- 'Use the following pydantic model json schema for each tool call you will make: {"properties": {"arguments": {"title": "Arguments", "type": "object"}, "name": {"title": "Name", "type": "string"}}, "required": ["arguments", "name"], "title": "FunctionCall", "type": "object"}
+' }}
+{{- "For each function call return a json object with function name and arguments within <tool_call></tool_call> XML tags as follows:
+" }}
+{{- "<tool_call>
+" }}
+{{- '{"arguments": <args-dict>, "name": <function-name>}
+' }}
+{{- '</tool_call><|im_end|>' }}
+{%- for message in messages %}
+    {%- if message.role == "user" or message.role == "system" or (message.role == "assistant" and message.tool_calls is not defined) %}
+        {{- '<|im_start|>' + message.role + '\n' + message.content + '<|im_end|>' + '\n' }}
+    {%- elif message.role == "assistant" %}
+        {{- '<|im_start|>' + message.role + '\n<tool_call>\n' }}
+        {%- for tool_call in message.tool_calls %}
+            {%- if tool_call.function is defined %}
+                {%- set tool_call = tool_call.function %}
+            {%- endif %}
+            {{- '{ ' }}
+            {%- if tool_call.arguments is defined %}
+                {{- '"arguments": ' }}
+                {{- tool_call.arguments|tojson }}
+                {{- ', '}}
+            {%- endif %}
+            {{- '"name": "' }}
+            {{- tool_call.name }}
+            {{- '"}' }}
+            {{- '\n</tool_call> ' }}
+        {%- endfor %}
+        {{- '<|im_end|>\n' }}
+    {%- elif message.role == "tool" %}
+        {%- if not message.name is defined %}
+            {{- raise_exception("Tool response dicts require a 'name' key indicating the name of the called function!") }}
```

Comment:
I don´t know why, but I had to remove this exception to make it work with Lllama-index's OpenAIAgent.

See: https://docs.llamaindex.ai/en/stable/examples/agent/openai_agent/

It may be a problem in Llama-index's implementation. I'm investigating more now.

Reply:
Yeah I was kind of stumped by this, I ran into some pydantic errors with the `name` property on tool use messages. I need to dig more into it - I think I remember that the OpenAI SDK's type specification didn't match up with the current spec [here](https://platform.openai.com/docs/api-reference/chat/create) - maybe an older version of the OpenAI SDK is specified in vLLM and the field has been deprecated? it seems like at least until recently, if not now still, `name` field was required on tool use messages

Reply:
Aha okay - yes it seems like OpenAI used to require it, and how they only require the tool call ID in `tool`-role messages. The Hermes 2 pro chat template does still require it, since it's how it tells the model which tool call the response pertains to - since otherwise you don't have that information when building the prompt from the template.

fwiw streaming tool calls is still not yet supported with Hermes models, although I'm working on it - so while non-streaming responses should work, streaming responses containing tools will probably be broken

Reply:
Found this related questions concerning the missing "name field" with Langchain and LLama-index:

 * https://community.openai.com/t/is-name-a-supported-parameter-for-tool-messages/843543
 * https://community.openai.com/t/is-name-a-supported-parameter-for-tool-messages/843543

Reply:
Couldn't the tool name be retrieved from the tool-call-id then ? I'm afraid many client will not send the name in the message as it is not required by OpenAI now (as you pointed out).

https://platform.openai.com/docs/api-reference/chat/create#chat-create-messages